mfr/extensions/zip/render.py

```diff
@@ -1,36 +1,188 @@
 import os
-import zipfile
+from typing import List, Union
+from zipfile import ZipFile, ZipInfo
 
-import markupsafe
 from mako.lookup import TemplateLookup
 
-from mfr.core import extension
 from mfr.core.utils import sizeof_fmt
+from mfr.core.extension import BaseRenderer
 
 
-class ZipRenderer(extension.BaseRenderer):
+class ZipRenderer(BaseRenderer):
 
     TEMPLATE = TemplateLookup(
         directories=[
             os.path.join(os.path.dirname(__file__), 'templates')
         ]).get_template('viewer.mako')
 
-    def render(self):
-        zip_file = zipfile.ZipFile(self.file_path, 'r')
-
-        filelist = [{'name': markupsafe.escape(file.filename),
-                     'size': sizeof_fmt(int(file.file_size)),
-                     'date': "%d-%02d-%02d %02d:%02d:%02d" % file.date_time[:6]} for file in zip_file.filelist
-                    if not file.filename.startswith('__MACOSX')]
-
-        message = '' if filelist else 'This zip file is empty.'
-
-        return self.TEMPLATE.render(zipped_filenames=filelist, message=message)
-
     @property
     def file_required(self):
         return True
 
     @property
     def cache_result(self):
         return True
+
+    def render(self):
+
+        zip_file = ZipFile(self.file_path, 'r')
+
+        # ``ZipFile.filelist`` contains both files and folder.  Using ``obj`` for better clarity.
+        obj_list = self.sanitize_obj_list(zip_file.filelist)
+        obj_tree = self.obj_list_to_tree(obj_list)
+
+        return self.TEMPLATE.render(data=obj_tree, base=self.assets_url)
+
+    def obj_list_to_tree(self, obj_list: list) -> List[dict]:
```

Comment:
Please take a look at the comments in the test fixtures for the structure of `obj_list` and `obj_tree`. Hopefully the in-code comments in this method provides enough information.

I'd like to mention one quirk in the tree building process. Given an object list of ['A/ab.png', 'A/', ], the algorithm first sees object 'A/ab.png'. It tries to build the tree by adding node `A/` to the root's children list and then adding `ab.png` to `A`'s children list. However, at this time we don't have details (`icon`, `data`) of `A/`. The solution is to build the structure only in this iteration and update the details in later iterations when  `A/` is the object.

Reply:
Add an axample of fixture to the docstring?